### ISSUE TITLE
[Bug] The setup is executed twice in the crud controller

### ISSUE THREAD

**Issue by DanielKimmich** [Laravel-Backpack/CRUD#3507]
# Bug report
The setup is executed twice in the crud controller, generating duplicate information

![image](https://user-images.githubusercontent.com/44840413/106741384-5c516680-65fa-11eb-9b9e-b9511b974f2e.png)

### What I did

add the "InlineCreate" operation and from there the setup function is repeated. As seen in the image.

```
class ContactPersonCrudController extends CrudController
{
    use \Backpack\CRUD\app\Http\Controllers\Operations\ListOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\ShowOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation { store as traitStore; }
    use \Backpack\CRUD\app\Http\Controllers\Operations\InlineCreateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation { update as traitUpdate; }
    use \Backpack\CRUD\app\Http\Controllers\Operations\DeleteOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\FetchOperation;

      public function setup()
    {
        $this->crud->setModel('App\Models\ContactPerson');
        $this->crud->setRoute(config('backpack.base.route_prefix') . '/contactperson');
        $this->crud->setEntityNameStrings(trans('contact.person.entity_name'), trans('contact.person.entity_names'));
        $this->setAccessOperation('contactperson');

    // ------ CRUD DETAILS ROW
        $this->crud->enableDetailsRow();
        //$this->crud->disableDetailsRow();
        $this->crud->setDetailsRowView('vendor.backpack.crud.details_row_contact');

        // display lead status counts on page top
        $contactPersonCount = ContactPerson::count();
            Widget::add([
                'type'        => 'progress',
                'class'       => 'card text-white bg-success mb-2',
                'value'       => $contactPersonCount,
                'description' => trans('dashboard.dataprogress.new_person'), //'nuevas Personas',
                'progress'    => $contactPersonCount, // integer
                'hint'        => trans('dashboard.dataprogress.hint'),
            ])->to('before_content');
            
    }

```

### What I expected to happen

that the "Widget" is only displayed once

### What I've already tried to fix it

If I comment on the "InlineCreate" operation that doesn't happen

` //   use \Backpack\CRUD\app\Http\Controllers\Operations\InlineCreateOperation;`

### PHP VERSION:
PHP 7.4.7 (cli) (built: Jun  9 2020 13:36:15) ( ZTS Visual C++ 2017 x64 )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies

### LARAVEL VERSION:
v7.30.4@9dd38140dc2924daa1a020a3d7a45f9ceff03df3

### BACKPACK VERSION:
4.1.32@1a65e15562989c8fda3d2941da7110ca4008e21e

**Comment by pxpm** [comment]
Hello @DanielKimmich 

Indeed, `setup()` is not run only one time. If you remove `InlineCreate` from a CRUD controller, and in setup you `Log::info('monster setup called.')` you will endup with:

```php
[2021-02-06 12:24:46] local.INFO: monster setup called.  
[2021-02-06 12:24:47] local.INFO: monster setup called.
```

`setup()` is called inside middleware in `CrudController` 

```php
public function __construct()
    {

        if ($this->crud) {
            return;
        }

        // ---------------------------
        // Create the CrudPanel object
        // ---------------------------
        // Used by developers inside their ProductCrudControllers as
        // $this->crud or using the CRUD facade.
        //
        // It's done inside a middleware closure in order to have
        // the complete request inside the CrudPanel object.
        $this->middleware(function ($request, $next) {
            $this->crud = app()->make('crud');
            $this->crud->setRequest($request);

            $this->setupDefaults();
            Log::info('running setup inside middleware closure');
            $this->setup();
            Log::info('runned setup inside middleware closure');
            $this->setupConfigurationForCurrentOperation();

            return $next($request);
        });
    }
```

You will endup with: 

```php
[2021-02-06 13:39:04] local.INFO: running setup inside middleware closure  
[2021-02-06 13:39:04] local.INFO: runned setup inside middleware closure  
[2021-02-06 13:39:05] local.INFO: running setup inside middleware closure  
[2021-02-06 13:39:05] local.INFO: runned setup inside middleware closure  
```

While using InlineCreateOperation setup will be called one more time. That is needed because you define fields in `setupCreateOperation` but also some people define them in `setup()`  and we need to fetch them to use in InlineCreateOperation. 

I would advise you to not use `setup()` for anything that is OPERATION specific. If you want to use something in more than one operation you can create a function and call it in `setupSomeOperation` or in all the operations you want. 

Just move your widget to be used in the operation you want, and you should be fine. 

I will be closing this, if you think I am wrong, and sometimes I am, please re-open. 

Wish you the best,
Pedro

**Comment by DanielKimmich** [comment]
Thank you very much. I will move it to "setupListOperation ()"

**Comment by tabacitu** [comment]
Hmmm... @pxpm I didn't know `setup()` gets called twice using `InlineCreate`... I agree with @DanielKimmich here - it sounds a bit unexpected... I don't think it's a HUGE problem, I don't expect people to do destructive actions inside `setup()`, but... yeah, if they do things like `addColumn()` or `addField()` or `addFilter()`, it'll result in unexpected behaviour for them. Especially since no other operation works like that... 

I'm wondering... any way around this? Do we _really need_ to call `setup()` inside `setupInlineCreateDefaults()`?

To sum it up, by now the _quirks` of InlineCreate are that:
- it needs to be used _after_ the Create operation in order to work;
- your `setup()` will be called twice;

**Comment by pxpm** [comment]
@tabacitu I'v just submitted PR #3521 

It fixes both `quirks`, but not without a tough call! Check it out. 

Pedro
